From 129b7e402bd6e7278854e5a8935fce460552b5f4 Mon Sep 17 00:00:00 2001
From: Thomas Markwalder <tmark@isc.org>
Date: Thu, 30 Jul 2020 10:01:36 -0400
Subject: [PATCH] [#117] Fixed gcc 10 compilation issues

client/dhclient.c
relay/dhcrelay.c
    extern'ed local_port,remote_port

common/discover.c
    init local_port,remote_port to 0

server/mdb.c
    extern'ed dhcp_type_host

server/mdb6.c
    create_prefix6() - eliminated memcpy string overflow error
---
 RELNOTES          | 5 +++++
 client/dhclient.c | 5 +++--
 common/discover.c | 4 ++--
 relay/dhcrelay.c  | 4 ++--
 server/mdb.c      | 2 +-
 server/mdb6.c     | 2 +-
 6 files changed, 14 insertions(+), 8 deletions(-)

#diff --git a/RELNOTES b/RELNOTES
#index 9d0a0414..6919dba7 100644
#--- a/RELNOTES
#+++ b/RELNOTES
#@@ -103,6 +103,11 @@ ISC DHCP is open source software maintained by Internet Systems
# Consortium.  This product includes cryptographic software written
# by Eric Young (eay@cryptsoft.com).
# 
#+		Changes since 4.4.2 (Bug Fixes)
#+
#+- Minor corrections to allow compilation under gcc 10.
#+  [Gitlab #117]
#+
# 		Changes since 4.4.2b1 (Bug Fixes)
# 
# - Added a clarification on DHCPINFORMs and server authority to
Index: isc-dhcp-4.4.1/client/dhclient.c
===================================================================
--- isc-dhcp-4.4.1.orig/client/dhclient.c
+++ isc-dhcp-4.4.1/client/dhclient.c
@@ -81,8 +81,9 @@ static const char message [] = "Internet
 static const char url [] = "For info, please visit https://www.isc.org/software/dhcp/";
 #endif /* UNIT_TEST */
 
-u_int16_t local_port = 0;
-u_int16_t remote_port = 0;
+extern u_int16_t local_port;
+extern u_int16_t remote_port;
+
 #if defined(DHCPv6) && defined(DHCP4o6)
 int dhcp4o6_state = -1; /* -1 = stopped, 0 = polling, 1 = started */
 #endif
Index: isc-dhcp-4.4.1/relay/dhcrelay.c
===================================================================
--- isc-dhcp-4.4.1.orig/relay/dhcrelay.c
+++ isc-dhcp-4.4.1/relay/dhcrelay.c
@@ -95,8 +95,8 @@ enum { forward_and_append,	/* Forward an
        forward_untouched,	/* Forward without changes. */
        discard } agent_relay_mode = forward_and_replace;
 
-u_int16_t local_port;
-u_int16_t remote_port;
+extern u_int16_t local_port;
+extern u_int16_t remote_port;
 
 /* Relay agent server list. */
 struct server_list {
Index: isc-dhcp-4.4.1/server/mdb.c
===================================================================
--- isc-dhcp-4.4.1.orig/server/mdb.c
+++ isc-dhcp-4.4.1/server/mdb.c
@@ -67,7 +67,7 @@ static host_id_info_t *host_id_info = NU
 
 int numclasseswritten;
 
-omapi_object_type_t *dhcp_type_host;
+extern omapi_object_type_t *dhcp_type_host;
 
 isc_result_t enter_class(cd, dynamicp, commit)
 	struct class *cd;
Index: isc-dhcp-4.4.1/server/mdb6.c
===================================================================
--- isc-dhcp-4.4.1.orig/server/mdb6.c
+++ isc-dhcp-4.4.1/server/mdb6.c
@@ -1943,7 +1943,7 @@ create_prefix6(struct ipv6_pool *pool, s
 		}
 		new_ds.data = new_ds.buffer->data;
 		memcpy(new_ds.buffer->data, ds.data, ds.len);
-		memcpy(new_ds.buffer->data + ds.len, &tmp, sizeof(tmp));
+		memcpy(&new_ds.buffer->data[0] + ds.len, &tmp, sizeof(tmp));
 		data_string_forget(&ds, MDL);
 		data_string_copy(&ds, &new_ds, MDL);
 		data_string_forget(&new_ds, MDL);
